cpp/src/arrow/util/decimal_test.cc

```diff
@@ -1119,6 +1137,35 @@ class TestDecimalToReal : public ::testing::Test {
     // 2**64 + 2**41 (exactly representable in a float)
     CheckDecimalToReal<Decimal, Real>("18446746272732807168", 0, 1.8446746e+19f);
     CheckDecimalToReal<Decimal, Real>("-18446746272732807168", 0, -1.8446746e+19f);
+
+    // Integers are always exact
+    auto scale = Decimal::kMaxScale - 1;
+    std::string seven = "7.";
+    for (int32_t i = 0; i < scale; ++i) {
+      seven += "0";
+    }
+    CheckDecimalToReal<Decimal, Real>(seven, scale, 7.0f);
+    CheckDecimalToReal<Decimal, Real>("-" + seven, scale, -7.0f);
+
+    CheckDecimalToReal<Decimal, Real>("99999999999999999999.0000000000000000", 16,
+                                      99999999999999999999.0f);
+    CheckDecimalToReal<Decimal, Real>("-99999999999999999999.0000000000000000", 16,
+                                      -99999999999999999999.0f);
+
+    // Small fractions are within one ULP
+    CheckDecimalToRealWithinOneULP<Decimal, Real>("9999999.9", 1, 9999999.9f);
+    CheckDecimalToRealWithinOneULP<Decimal, Real>("-9999999.9", 1, -9999999.9f);
+    CheckDecimalToRealWithinOneULP<Decimal, Real>("9999999.999999", 6, 9999999.999999f);
+    CheckDecimalToRealWithinOneULP<Decimal, Real>("-9999999.999999", 6, -9999999.999999f);
+
+    // Large fractions are within 2^-23
+    constexpr Real epsilon = 1.1920928955078125e-07f;  // 2^-23
+    CheckDecimalToRealWithinEpsilon<Decimal, Real>(
+        "112334829348925.99070703983306884765625", 23, epsilon,
+        112334829348925.99070703983306884765625f);
```

Comment:
Passing this as a `float` constant is weird. If you want to differentiate these tests between `float` and `double`, you may use `if constexpr`.

Reply:
The weird thing is even though it is parametrized over `Real`, these tests were only run for `float`. I kept it this way. Perhaps it's better to remove the Real parameter and make it clear that it's only for float?

Reply:
Either that, or try to re-enable running them for `double`. Are the tests redundant with other `double` tests?

Reply:
Yes, `double` is tested in TestDecimalToRealDouble::Precision. I added a static_assert at the beginning to make it clearer.

Reply:
If the test parameterization is fake, then let's just remove it?

Reply:
Sure. I've refactored this test as TestDecimalToRealFloat.